<a href="https://colab.research.google.com/github/Natali1599/MineriaDatos/blob/main/Copia_de_Parcial1MineriaDatos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Parcial 1 Minería de datos

In [1]:
import pandas as pd
import numpy as np
import csv

In [2]:
pd.set_option('display.max_columns', None) #Mostrar todas las columnas


In [85]:
df = pd.read_csv(
    'https://raw.githubusercontent.com/prof-tatiana-gutierrez/DatosMD/master/student-mat.csv',
    sep = ";")
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  
0     1       3         6   5   6   6  
1     1       3         4   5   5   6  
2     3       3        10   7   8  10  
3     1       5         2  15  14  15  
4     2       5         4   6  10  10

In [64]:
df.shape

(395, 33)

In [4]:
df.isna().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

In [5]:
df.isnull().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

In [6]:
pd.unique(df['address'])

array(['U', 'R'], dtype=object)

### 1. Valores vacios o nulos

In [12]:
df.isna()

school    sex    age  address  famsize  Pstatus   Medu   Fedu   Mjob  \
0     False  False  False    False    False    False  False  False  False   
1     False  False  False    False    False    False  False  False  False   
2     False  False  False    False    False    False  False  False  False   
3     False  False  False    False    False    False  False  False  False   
4     False  False  False    False    False    False  False  False  False   
..      ...    ...    ...      ...      ...      ...    ...    ...    ...   
390   False  False  False    False    False    False  False  False  False   
391   False  False  False    False    False    False  False  False  False   
392   False  False  False    False    False    False  False  False  False   
393   False  False  False    False    False    False  False  False  False   
394   False  False  False    False    False    False  False  False  False   

      Fjob  reason  guardian  traveltime  studytime  failures  schoolsup  \
0    False   False     False       False      False     False      False   
1    False   False     False       False      False     False      False   
2    False   False     False       False      False     False      False   
3    False   False     False       False      False     False      False   
4    False   False     False       False      False     False      False   
..     ...     ...       ...         ...        ...       ...        ...   
390  False   False     False       False      False     False      False   
391  False   False     False       False      False     False      False   
392  False   False     False       False      False     False      False   
393  False   False     False       False      False     False      False   
394  False   False     False       False      False     False      False   

     famsup   paid  activities  nursery  higher  internet  romantic  famrel  \
0     False  False       False    False   False     False     False   False   
1     False  False       False    False   False     False     False   False   
2     False  False       False    False   False     False     False   False   
3     False  False       False    False   False     False     False   False   
4     False  False       False    False   False     False     False   False   
..      ...    ...         ...      ...     ...       ...       ...     ...   
390   False  False       False    False   False     False     False   False   
391   False  False       False    False   False     False     False   False   
392   False  False       False    False   False     False     False   False   
393   False  False       False    False   False     False     False   False   
394   False  False       False    False   False     False     False   False   

     freetime  goout   Dalc   Walc  health  absences     G1     G2     G3  
0       False  False  False  False   False     False  False  False  False  
1       False  False  False  False   False     False  False  False  False  
2       False  False  False  False   False     False  False  False  False  
3       False  False  False  False   False     False  False  False  False  
4       False  False  False  False   False     False  False  False  False  
..        ...    ...    ...    ...     ...       ...    ...    ...    ...  
390     False  False  False  False   False     False  False  False  False  
391     False  False  False  False   False     False  False  False  False  
392     False  False  False  False   False     False  False  False  False  
393     False  False  False  False   False     False  False  False  False  
394     False  False  False  False   False     False  False  False  False  

[395 rows x 33 columns]

No tiene datos vacios o nulos.

### 2. ¿cuántos estudiantes viven en zona urbana o rural? ¿cuántos de los estudiante rurales son mujeres? 

In [15]:
df["address"].value_counts()

U    307
R     88
Name: address, dtype: int64

En la zona urbana viven 307 y en la zona rural viven 88

In [16]:
df["sex"].value_counts()

F    208
M    187
Name: sex, dtype: int64

In [17]:
tab1 = pd.crosstab(df["address"],df["sex"])
tab1

sex        F    M
address          
R         44   44
U        164  143

Mujeres estudiantes rurales son 44.

### 3. ¿cuántos estudiantes que viven en la zona rural tiene menos de 5 fallas? ¿Cuántos de ellos son mujeres?

In [20]:
df["absences"].value_counts()

0     115
2      65
4      53
6      31
8      22
10     17
14     12
12     12
3       8
16      7
7       7
5       5
18      5
20      4
11      3
9       3
13      3
15      3
22      3
1       3
38      1
30      1
40      1
23      1
19      1
28      1
75      1
21      1
24      1
56      1
26      1
54      1
25      1
17      1
Name: absences, dtype: int64

In [40]:
dfrural = df[df['address']=="R"]
dfrural

school sex  age address famsize Pstatus  Medu  Fedu      Mjob     Fjob  \
24      GP   F   15       R     GT3       T     2     4  services   health   
32      GP   M   15       R     GT3       T     4     3   teacher  at_home   
37      GP   M   16       R     GT3       A     4     4     other  teacher   
38      GP   F   15       R     GT3       T     3     4  services   health   
39      GP   F   15       R     GT3       T     2     2   at_home    other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...      ...   
385     MS   F   18       R     GT3       T     2     2   at_home    other   
386     MS   F   18       R     GT3       T     4     4   teacher  at_home   
387     MS   F   19       R     GT3       T     2     3  services    other   
392     MS   M   21       R     GT3       T     1     1     other    other   
393     MS   M   18       R     LE3       T     3     2  services    other   

         reason guardian  traveltime  studytime  failures schoolsup famsup  \
24       course   mother           1          3         0       yes    yes   
32       course   mother           1          2         0        no    yes   
37   reputation   mother           2          3         0        no    yes   
38       course   mother           1          3         0       yes    yes   
39   reputation   mother           1          1         0       yes    yes   
..          ...      ...         ...        ...       ...       ...    ...   
385       other   mother           2          3         0        no     no   
386  reputation   mother           3          1         0        no    yes   
387      course   mother           1          3         1        no     no   
392      course    other           1          1         3        no     no   
393      course   mother           3          1         0        no     no   

    paid activities nursery higher internet romantic  famrel  freetime  goout  \
24   yes        yes     yes    yes      yes       no       4         3      2   
32    no        yes     yes    yes      yes      yes       4         5      2   
37    no        yes     yes    yes      yes      yes       2         4      3   
38   yes        yes     yes    yes      yes       no       4         3      2   
39   yes        yes     yes    yes       no       no       4         3      1   
..   ...        ...     ...    ...      ...      ...     ...       ...    ...   
385  yes         no     yes    yes       no       no       5         3      3   
386  yes        yes     yes    yes      yes      yes       4         4      3   
387   no        yes      no    yes      yes       no       5         4      2   
392   no         no      no    yes       no       no       5         5      3   
393   no         no      no    yes      yes       no       4         4      1   

     Dalc  Walc  health  absences  G1  G2  G3  
24      1     1       5         2  10   9   8  
32      1     1       5         0  17  16  16  
37      1     1       5         7  15  16  15  
38      1     1       5         2  12  12  11  
39      1     1       2         8  14  13  13  
..    ...   ...     ...       ...  ..  ..  ..  
385     1     3       4         2  10   9  10  
386     2     2       5         7   6   5   6  
387     1     2       5         0   7   5   0  
392     3     3       3         3  10   8   7  
393     3     4       5         0  11  12  10  

[88 rows x 33 columns]

In [44]:
tab4=dfrural[dfrural['absences'] < 5]
tab4

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
24      GP   F   15       R     GT3       T     2     4  services    health   
32      GP   M   15       R     GT3       T     4     3   teacher   at_home   
38      GP   F   15       R     GT3       T     3     4  services    health   
68      GP   F   15       R     LE3       T     2     2    health  services   
72      GP   F   15       R     GT3       T     1     1     other     other   
95      GP   F   15       R     GT3       T     1     1   at_home     other   
96      GP   M   16       R     GT3       T     4     3  services     other   
111     GP   F   16       R     GT3       T     3     3  services     other   
128     GP   M   18       R     GT3       T     2     2  services     other   
130     GP   F   15       R     GT3       T     3     4  services   teacher   
134     GP   M   15       R     GT3       T     3     4   at_home   teacher   
136     GP   M   17       R     GT3       T     3     4   at_home     other   
155     GP   M   15       R     GT3       T     2     3   at_home  services   
158     GP   M   16       R     GT3       T     2     2   at_home     other   
160     GP   M   17       R     LE3       T     2     1   at_home     other   
164     GP   M   17       R     LE3       T     1     1     other  services   
189     GP   M   17       R     GT3       T     1     2   at_home     other   
220     GP   F   17       R     GT3       T     2     1   at_home  services   
231     GP   M   17       R     GT3       T     2     2     other     other   
238     GP   F   17       R     GT3       T     2     1   at_home  services   
241     GP   M   17       R     LE3       A     4     4   teacher     other   
253     GP   M   16       R     GT3       T     2     1     other     other   
254     GP   M   17       R     GT3       T     2     1     other     other   
262     GP   M   18       R     GT3       T     3     2     other     other   
269     GP   F   18       R     GT3       T     2     1     other     other   
273     GP   M   17       R     GT3       T     1     2   at_home   at_home   
282     GP   F   18       R     LE3       T     1     1   at_home     other   
308     GP   M   19       R     GT3       T     3     3     other  services   
318     GP   F   17       R     GT3       T     3     4   at_home  services   
322     GP   F   17       R     LE3       T     2     2  services  services   
334     GP   F   18       R     GT3       T     2     2   at_home     other   
339     GP   F   17       R     GT3       A     3     2     other     other   
353     MS   M   19       R     GT3       T     1     1     other     other   
354     MS   M   17       R     GT3       T     4     3  services     other   
356     MS   F   17       R     GT3       T     4     4   teacher  services   
360     MS   F   18       R     LE3       A     1     4   at_home     other   
361     MS   M   18       R     LE3       T     1     1   at_home     other   
364     MS   F   17       R     GT3       T     1     2     other  services   
365     MS   M   18       R     GT3       T     1     3   at_home     other   
367     MS   F   17       R     GT3       T     1     1     other  services   
371     MS   M   18       R     LE3       T     1     2   at_home  services   
374     MS   F   18       R     LE3       T     4     4     other     other   
375     MS   F   18       R     GT3       T     1     1     other     other   
377     MS   F   18       R     LE3       T     4     4   teacher  services   
383     MS   M   19       R     GT3       T     1     1     other  services   
385     MS   F   18       R     GT3       T     2     2   at_home     other   
387     MS   F   19       R     GT3       T     2     3  services     other   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   

         reason guardian  traveltime  studytime  fail

In [55]:
tab4.shape

(49, 33)

La cantidad de estudiantes que viven en la zona rural y tienen menos de 5 fallas son 49.

In [51]:
tab5=tab4[tab4['sex']=="M"] 
tab5

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
32      GP   M   15       R     GT3       T     4     3   teacher   at_home   
96      GP   M   16       R     GT3       T     4     3  services     other   
128     GP   M   18       R     GT3       T     2     2  services     other   
134     GP   M   15       R     GT3       T     3     4   at_home   teacher   
136     GP   M   17       R     GT3       T     3     4   at_home     other   
155     GP   M   15       R     GT3       T     2     3   at_home  services   
158     GP   M   16       R     GT3       T     2     2   at_home     other   
160     GP   M   17       R     LE3       T     2     1   at_home     other   
164     GP   M   17       R     LE3       T     1     1     other  services   
189     GP   M   17       R     GT3       T     1     2   at_home     other   
231     GP   M   17       R     GT3       T     2     2     other     other   
241     GP   M   17       R     LE3       A     4     4   teacher     other   
253     GP   M   16       R     GT3       T     2     1     other     other   
254     GP   M   17       R     GT3       T     2     1     other     other   
262     GP   M   18       R     GT3       T     3     2     other     other   
273     GP   M   17       R     GT3       T     1     2   at_home   at_home   
308     GP   M   19       R     GT3       T     3     3     other  services   
353     MS   M   19       R     GT3       T     1     1     other     other   
354     MS   M   17       R     GT3       T     4     3  services     other   
361     MS   M   18       R     LE3       T     1     1   at_home     other   
365     MS   M   18       R     GT3       T     1     3   at_home     other   
371     MS   M   18       R     LE3       T     1     2   at_home  services   
383     MS   M   19       R     GT3       T     1     1     other  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   

         reason guardian  traveltime  studytime  failures schoolsup famsup  \
32       course   mother           1          2         0        no    yes   
96   reputation   mother           2          1         0       yes    yes   
128  reputation   mother           1          1         2        no    yes   
134      course   mother           4          2         0        no    yes   
136      course   mother           3          2         0        no     no   
155      course   mother           1          2         0       yes     no   
158      course   mother           3          1         0        no     no   
160      course   mother           2          1         2        no     no   
164      course   mother           4          2         3        no     no   
189        home   mother           1          2         0        no     no   
231      course   father           2          2         0        no    yes   
241      course   mother           2          2         0        no    yes   
253      course   mother           2          1         0        no     no   
254      course   mother           1          1         0        no     no   
262      course   mother           1          3         0        no     no   
273        home   mother           1          2         0        no    yes   
308  reputation   father           1          2         1        no     no   
353        home    other           3          1         1        no    yes   
354        home   mother           2          2         0        no    yes   
361       other   mother           2          2         1        no     no   
365      course   mother           2          2         0        no    yes   
371       other   father           3          1         0        no    yes   
383       other   mother           2          1         1        no     no   
392      course    other           1          1         3        no     no   

In [54]:
tab5.shape

(25, 33)

La cantidad de mujeres de la zona rural que tienen menos de 5 ausencias son 25.

### 4. Ver la media de las notas de matemáticas de primer corte (G1) de los estudiantes deel colegio gabriel pereira vs Mousinho da Silveita, teniendo en cuenta si los estudiantes tienen o no internet en sus casas. 

In [60]:
colgp = df[df['school']=="GP"]
colgp

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4  at_home   teacher   
1       GP   F   17       U     GT3       T     1     1  at_home     other   
2       GP   F   15       U     LE3       T     1     1  at_home     other   
3       GP   F   15       U     GT3       T     4     2   health  services   
4       GP   F   16       U     GT3       T     3     3    other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...      ...       ...   
344     GP   F   18       U     GT3       T     2     3  at_home     other   
345     GP   F   18       U     GT3       T     3     2    other  services   
346     GP   M   18       R     GT3       T     4     3  teacher  services   
347     GP   M   18       U     GT3       T     4     3  teacher     other   
348     GP   F   17       U     GT3       T     4     3   health     other   

         reason guardian  traveltime  studytime  failures schoolsup famsup  \
0        course   mother           2          2         0       yes     no   
1        course   father           1          2         0        no    yes   
2         other   mother           1          2         3       yes     no   
3          home   mother           1          3         0        no    yes   
4          home   father           1          2         0        no    yes   
..          ...      ...         ...        ...       ...       ...    ...   
344      course   mother           1          3         0        no    yes   
345       other   mother           1          3         0        no     no   
346      course   mother           1          3         0        no     no   
347      course   mother           1          3         0        no    yes   
348  reputation   mother           1          3         0        no    yes   

    paid activities nursery higher internet romantic  famrel  freetime  goout  \
0     no         no     yes    yes       no       no       4         3      4   
1     no         no      no    yes      yes       no       5         3      3   
2    yes         no     yes    yes      yes       no       4         3      2   
3    yes        yes     yes    yes      yes      yes       3         2      2   
4    yes         no     yes    yes       no       no       4         3      2   
..   ...        ...     ...    ...      ...      ...     ...       ...    ...   
344   no         no     yes    yes      yes       no       4         3      3   
345   no         no     yes    yes      yes      yes       5         4      3   
346   no         no     yes    yes      yes      yes       5         3      2   
347  yes         no     yes    yes      yes      yes       5         4      5   
348  yes        yes     yes    yes      yes      yes       4         4      3   

     Dalc  Walc  health  absences  G1  G2  G3  
0       1     1       3         6   5   6   6  
1       1     1       3         4   5   5   6  
2       2     3       3        10   7   8  10  
3       1     1       5         2  15  14  15  
4       1     2       5         4   6  10  10  
..    ...   ...     ...       ...  ..  ..  ..  
344     1     2       3         4  11  10  10  
345     2     3       1         7  13  13  14  
346     1     2       4         9  16  15  16  
347     2     3       5         0  10  10   9  
348     1     3       4         0  13  15  15  

[349 rows x 33 columns]

In [61]:
colgp.shape

(349, 33)

In [62]:
colms = df[df['school']=="MS"]
colms

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
349     MS   M   18       R     GT3       T     3     2     other     other   
350     MS   M   19       R     GT3       T     1     1     other  services   
351     MS   M   17       U     GT3       T     3     3    health     other   
352     MS   M   18       U     LE3       T     1     3   at_home  services   
353     MS   M   19       R     GT3       T     1     1     other     other   
354     MS   M   17       R     GT3       T     4     3  services     other   
355     MS   F   18       U     GT3       T     3     3  services  services   
356     MS   F   17       R     GT3       T     4     4   teacher  services   
357     MS   F   17       U     LE3       A     3     2  services     other   
358     MS   M   18       U     LE3       T     1     1     other  services   
359     MS   F   18       U     LE3       T     1     1   at_home  services   
360     MS   F   18       R     LE3       A     1     4   at_home     other   
361     MS   M   18       R     LE3       T     1     1   at_home     other   
362     MS   F   18       U     GT3       T     3     3  services  services   
363     MS   F   17       U     LE3       T     4     4   at_home   at_home   
364     MS   F   17       R     GT3       T     1     2     other  services   
365     MS   M   18       R     GT3       T     1     3   at_home     other   
366     MS   M   18       U     LE3       T     4     4   teacher  services   
367     MS   F   17       R     GT3       T     1     1     other  services   
368     MS   F   18       U     GT3       T     2     3   at_home  services   
369     MS   F   18       R     GT3       T     4     4     other   teacher   
370     MS   F   19       U     LE3       T     3     2  services  services   
371     MS   M   18       R     LE3       T     1     2   at_home  services   
372     MS   F   17       U     GT3       T     2     2     other   at_home   
373     MS   F   17       R     GT3       T     1     2     other     other   
374     MS   F   18       R     LE3       T     4     4     other     other   
375     MS   F   18       R     GT3       T     1     1     other     other   
376     MS   F   20       U     GT3       T     4     2    health     other   
377     MS   F   18       R     LE3       T     4     4   teacher  services   
378     MS   F   18       U     GT3       T     3     3     other     other   
379     MS   F   17       R     GT3       T     3     1   at_home     other   
380     MS   M   18       U     GT3       T     4     4   teacher   teacher   
381     MS   M   18       R     GT3       T     2     1     other     other   
382     MS   M   17       U     GT3       T     2     3     other  services   
383     MS   M   19       R     GT3       T     1     1     other  services   
384     MS   M   18       R     GT3       T     4     2     other     other   
385     MS   F   18       R     GT3       T     2     2   at_home     other   
386     MS   F   18       R     GT3       T     4     4   teacher   at_home   
387     MS   F   19       R     GT3       T     2     3  services     other   
388     MS   F   18       U     LE3       T     3     1   teacher  services   
389     MS   F   18       U     GT3       T     1     1     other     other   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

         reason guardian  traveltime  studytime  failures schoolsup famsup  \
349      course   mother           2          1         1        no    yes   
350        home    other           3          2         3        no     no   
351      course   mother           2          2         

In [63]:
colms.shape

(46, 33)

In [94]:
#internet
#G1
tabgg=pd.pivot_table(colms, index=['G1','internet'], aggfunc=np.mean)
tabgg

Dalc      Fedu         G2         G3      Medu      Walc  \
G1 internet                                                                 
6  no        2.000000  1.333333   5.000000   1.666667  2.000000  2.333333   
   yes       1.500000  3.000000   5.000000   5.500000  2.500000  2.500000   
7  no        1.000000  2.000000   7.000000   9.000000  3.000000  1.000000   
   yes       1.000000  1.500000   6.500000   3.750000  2.000000  2.000000   
8  yes       2.500000  1.833333   8.000000   8.833333  1.500000  3.000000   
9  no        4.000000  2.000000   9.000000   9.000000  2.000000  5.000000   
10 no        1.750000  2.250000   9.000000   9.000000  1.750000  2.750000   
   yes       1.500000  1.000000  10.000000  10.000000  2.000000  2.500000   
11 yes       2.200000  2.600000  11.400000  10.600000  2.600000  3.000000   
12 no        1.000000  2.000000  11.500000  11.500000  2.000000  2.000000   
   yes       1.000000  4.000000  13.000000  13.000000  4.000000  2.000000   
13 no        1.333333  2.333333  12.000000  12.000000  1.333333  2.666667   
   yes       1.666667  3.333333  12.333333  12.333333  3.666667  2.666667   
14 yes       3.000000  2.333333  13.333333  13.000000  2.666667  3.000000   
15 yes       1.000000  3.000000  14.333333  14.666667  3.666667  2.333333   
16 yes       1.000000  4.000000  15.000000  15.000000  4.000000  1.000000   
18 yes       1.000000  1.000000  16.000000  16.000000  1.000000  1.000000   
19 yes       1.000000  4.000000  18.000000  19.000000  4.000000  1.000000   

              absences        age  failures    famrel  freetime     goout  \
G1 internet                                                                 
6  no         4.666667  18.333333  1.000000  3.333333  2.666667  2.000000   
   yes       10.500000  17.500000  0.000000  3.500000  4.500000  4.000000   
7  no         4.000000  19.000000  2.000000  3.000000  2.000000  2.000000   
   yes        1.250000  18.000000  0.500000  4.500000  3.250000  2.500000   
8  yes        5.000000  18.500000  0.833333  4.166667  3.333333  3.166667   
9  no        11.000000  20.000000  2.000000  5.000000  5.000000  4.000000   
10 no         2.250000  18.750000  0.750000  4.500000  3.500000  3.500000   
   yes       10.500000  17.500000  0.000000  3.500000  4.000000  3.000000   
11 yes        2.400000  17.800000  0.200000  3.800000  3.600000  2.800000   
12 no         1.000000  17.000000  0.000000  2.000000  2.000000  2.500000   
   yes        4.000000  17.000000  0.000000  4.000000  3.000000  3.000000   
13 no         3.333333  17.666667  0.333333  3.666667  3.666667  3.333333   
   yes        2.000000  17.333333  0.000000  4.000000  4.000000  3.666667   
14 yes        5.333333  17.666667  0.000000  3.000000  3.000000  3.333333   
15 yes        2.666667  18.666667  0.666667  4.000000  2.333333  3.333333   
16 yes        0.000000  17.000000  0.000000  2.000000  3.000000  4.000000   
18 yes        0.000000  18.000000  0.000000  5.000000  3.000000  2.000000   
19 yes        0.000000  18.000000  0.000000  5.000000  4.000000  4.000000   

               health  studytime  traveltime  
G1 internet                                   
6  no        4.333333   1.333333    2.000000  
   yes       3.000000   1.000000    2.000000  
7  no        3.000000   2.000000    2.000000  
   yes       3.000000   1.750000    1.750000  
8  yes       3.500000   1.666667    2.166667  
9  no        4.000000   2.000000    1.000000  
10 no        3.750000   2.000000    1.500000  
   yes       2.000000   1.500000    1.500000  
11 yes       4.000000   1.400000    2.200000  
12 no        4.000000   2.000000    2.000000  
   yes       5.000000   2.000000    2.000000  
13 no        4.333333   2.333333    2.000000  
   yes       3.333333   2.333333    2.000000  
14 yes       3.333333   1.333333    2.666667  
15 yes       2.000000   2.333333    1.333333  
16 yes       1.000000   2.000000    1.000000  
18 yes       4.000000   3.000000    2.000000  
19 yes       1.000000   3.000000    2.0000

In [76]:
pd.pivot_table(colgp, index=['G1','internet'], aggfunc=np.mean)

Dalc      Fedu         G2         G3      Medu      Walc  \
G1 internet                                                                 
3  yes       1.000000  3.000000   5.000000   5.000000  3.000000  3.000000   
4  yes       1.000000  3.000000   0.000000   0.000000  3.000000  1.000000   
5  no        1.000000  2.500000   7.000000   6.500000  2.500000  3.000000   
   yes       1.000000  2.000000   3.800000   3.600000  2.400000  2.000000   
6  no        1.000000  2.666667   6.666667   5.333333  2.666667  1.333333   
   yes       1.750000  2.062500   5.812500   4.312500  2.625000  3.312500   
7  no        1.500000  1.750000   5.250000   5.750000  1.500000  3.000000   
   yes       1.535714  2.357143   7.178571   5.714286  2.714286  2.035714   
8  no        1.222222  2.555556   8.555556   8.555556  2.444444  2.555556   
   yes       1.269231  2.038462   8.307692   7.807692  2.346154  2.076923   
9  no        1.200000  1.800000   7.000000   6.000000  1.400000  3.000000   
   yes       1.640000  2.560000   9.160000   8.440000  2.800000  2.400000   
10 no        1.400000  3.200000   8.000000   6.200000  3.000000  2.000000   
   yes       1.875000  2.825000  10.175000   9.625000  3.025000  2.650000   
11 no        1.333333  2.666667  10.500000  10.666667  2.833333  1.500000   
   yes       1.428571  2.678571  10.821429  10.785714  2.964286  2.178571   
12 no        1.750000  1.250000  12.000000  13.000000  2.250000  2.250000   
   yes       1.500000  2.642857  11.607143  11.821429  2.714286  2.392857   
13 no        1.000000  1.500000  11.500000  11.500000  2.500000  1.000000   
   yes       1.400000  2.640000  13.040000  13.000000  2.800000  2.160000   
14 no        1.000000  2.000000  13.250000  13.250000  2.500000  1.750000   
   yes       1.391304  3.043478  13.695652  14.000000  3.173913  2.086957   
15 no        1.666667  2.333333  13.666667  13.333333  2.666667  2.000000   
   yes       1.055556  2.388889  14.555556  14.888889  3.222222  1.888889   
16 no        2.500000  3.000000  14.500000  15.000000  1.500000  3.000000   
   yes       1.157895  2.947368  15.842105  16.210526  3.000000  2.105263   
17 no        1.000000  2.000000  16.500000  16.500000  2.500000  1.500000   
   yes       1.333333  3.333333  16.500000  17.000000  3.666667  2.166667   
18 no        1.000000  1.000000  18.000000  18.000000  3.000000  1.000000   
   yes       1.333333  2.833333  18.333333  18.500000  3.333333  2.166667   
19 yes       1.000000  3.500000  18.500000  19.000000  4.000000  1.000000   

             absences        age  failures    famrel  freetime     goout  \
G1 internet                                                                
3  yes       8.000000  18.000000  1.000000  4.000000  3.000000  3.000000   
4  yes       0.000000  16.000000  2.000000  4.000000  3.000000  2.000000   
5  no        3.000000  17.500000  1.500000  4.500000  3.000000  4.500000   
   yes       4.400000  16.800000  1.600000  4.000000  4.000000  3.800000   
6  no        3.333333  16.666667  0.333333  4.000000  2.333333  3.333333   
   yes       6.625000  17.312500  1.375000  3.625000  3.125000  4.437500   
7  no        2.250000  17.000000  1.000000  3.250000  3.500000  3.000000   
   yes       5.571429  16.178571  0.642857  3.928571  3.142857  3.107143   
8  no        5.555556  16.111111  0.000000  3.888889  2.777778  2.666667   
   yes       4.923077  16.384615  0.461538  4.153846  3.269231  3.038462   
9  no        3.200000  17.000000  0.800000  4.400000  2.800000  3.600000   
   yes       9.280000  16.680000  0.320000  3.880000  3.280000  3.040000   
10 no        4.000000  16.600000  0.000000  3.600000  3.400000  3.800000   
   yes       7.250000  16.375000  0.175000  3.975000  3.225000  3.425000   
11 no        4.000000  16.166667  0.166667  4.166667  3.500000  3.000000   
   yes       7.571429  16.428571  0.107143  3.892857  3.214286  3.107143   
12 no        6.250000  16.500000  0.500000  4.500000  3.500000  2.500000   
   yes       6.428571  16.64285

### 5. Cree una variable en la tabla que corresponda a la calificación total para cada uno de los estudiantes teniendo en cuenta lo siguiente: 


*   G1 (35%)
*   G2 (35%)
*   G3 (30%)



In [80]:
from scipy import stats
import numpy as np

In [81]:
!pip install dfply

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 612 kB 32.5 MB/s 


In [82]:
from sklearn import datasets
from dfply import *
import numpy as np

In [86]:
Base10 = (df >>
 mutate(
   total = (X.G1*0.35+X.G2*0.35+X.G3*0.30),
 ))
Base10.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  total  
0     1       3         6   5   6   6   5.65  
1     1       3         4   5   5   6   5.30  
2     3       3        10   7   8  10   8.25  
3     1       5         2  15  14  15  14.65  
4     2       5         4   6  10  10   8.60

In [84]:
Base10.shape

(395, 34)

In [88]:
Base10['total'].describe()

count    395.000000
mean      10.692532
std        3.662308
min        1.400000
25%        8.350000
50%       10.700000
75%       13.325000
max       19.300000
Name: total, dtype: float64



*   La media es 10.69
*   La mediana es 3.66
*   Mínima calificación es 1.4
*   La máxima calificación es 19.30



### 6. Responda

* ¿Qué tipo de análisis realizaría ¿Regresión o clasificación?? 
Respuesta: Clasificación, por tener mayor cantidad de variables discretas.
* ¿Que variable usaría para predecir o clasificar? 
Respuesta: La nota total / notas  
* ¿cuáles variables tiene la hipótesis que podrían servir para pronosticar o clasificar? 
Respuesta: La nota total, nota por cortes, sexo, edad, zona, tiempo de estudio, reprobaciones, schoolsup, internet, ausencias. 


### Bibliografía 
https://archive.ics.uci.edu/ml/datasets/student+performance